Add `DataColumnSidecarsByRoot ` req/resp protocol

jimmygchen opened the pull request:
## Issue Addressed

Part of #4983.

## Proposed Changes

Add implementation for `DataColumnSidecarsByRoot`.

Currently blobs are served from three sources, in the following order
1. data availability checker (before we have all the blocks and blobs)
2. early attester cache (before we import the block)
3. blobs db (after import)

I've implemented the equivalent for data columns, and it currently reuses the blobs db but with a different `DbColumn` to separate data columns from blobs. 

This PR does not cover the following:
1. Sending outbound requests
2. Handling of responses from peers

jimmygchen committed:
Add stub for `DataColumnsByRoot`

jimmygchen committed:
Merge branch 'das' into data-cols-by-root

# Conflicts:
#	consensus/types/src/chain_spec.rs

jimmygchen committed:
Add basic implementation of serving RPC data column from DA checker.

jimmygchen committed:
Merge branch 'das' into data-cols-by-root

jimmygchen committed:
Store data columns in early attester cache and blobs db.

jacobkaufmann commented on beacon_node/beacon_chain/src/block_verification_types.rs:
```diff
@@ -65,6 +68,9 @@ enum RpcBlockInner<E: EthSpec> {
     /// This variant is used with parent lookups and by-range responses. It should have all blobs
     /// ordered, all block roots matching, and the correct number of blobs for this block.
     BlockAndBlobs(Arc<SignedBeaconBlock<E>>, BlobSidecarList<E>),
+    /// This variant is used with parent lookups and by-range responses. It should have all blobs
```
```suggestion
    /// This variant is used with parent lookups and by-range responses. It should have all data columns
```

jacobkaufmann commented on beacon_node/beacon_chain/src/data_availability_checker.rs:
```diff
@@ -223,6 +235,21 @@ impl<T: BeaconChainTypes> DataAvailabilityChecker<T> {
             .put_kzg_verified_blobs(gossip_blob.block_root(), vec![gossip_blob.into_inner()])
     }
 
+    /// Check if we've cached other data columns for this block. If it completes a set and we also
```
```suggestion
    /// Check if we've cached other data columns for this block. If it satisfies the custody requirement and we also
```

jacobkaufmann commented on beacon_node/beacon_chain/src/data_availability_checker.rs:
```diff
@@ -306,30 +343,33 @@ impl<T: BeaconChainTypes> DataAvailabilityChecker<T> {
-                        None
-                    };
+                (maybe_blob_list, maybe_data_column_list) => {
+                    let (verified_blobs, verified_data_column) =
```
```suggestion
                    let (verified_blobs, verified_data_columns) =
```

jacobkaufmann commented on beacon_node/beacon_chain/src/data_availability_checker/availability_view.rs:
```diff
@@ -90,6 +116,42 @@ pub trait AvailabilityView<E: EthSpec> {
+    ///
+    /// Data columns are only inserted if:
+    /// 1. The data column entry at the index is empty and no block exists.
+    /// 2. The block exists and its commitment matches the blob's commitment.
```
```suggestion
    /// 2. The block exists and its commitments matches the data column's commitments.
```

eserilev commented on beacon_node/beacon_chain/src/block_verification_types.rs:
```diff
@@ -478,12 +494,13 @@ impl<E: EthSpec> AsBlock<E> for AvailableBlock<E> {
-        let inner = match blobs_opt {
-            None => RpcBlockInner::Block(block),
-            Some(blobs) => RpcBlockInner::BlockAndBlobs(block, blobs),
+        let inner = match (blobs_opt, data_columns_opt) {
+            (None, None) => RpcBlockInner::Block(block),
+            (Some(blobs), _) => RpcBlockInner::BlockAndBlobs(block, blobs),
+            (_, Some(data_columns)) => RpcBlockInner::BlockAndDataColumns(block, data_columns),
```
What happens here if we have both blobs and data_columns available? I guess we'd return a `BlockAndBlobs` variant? Would that imply that if blobs are avail, we'll just skip the data sampling checks?

would a `BlockAndBlobsAndDataColumns` variant allow us to support both `blobs` and `data_columns` until were ready to migrate away from blob subnets completely?

jimmygchen replied:
Yeah, so originally I was thinking to prioritise blobs, but then it's kind of pointless because we aren't executing the data column code paths.

I've been thinking about this and talking to Sean about this as well. A potentially way to introduce this and test it out may be add column subnets and sampling (without making it mandatory for `is_data_avaialble`, perhaps log warnings if it fails), and keep the existing blob subnets and DA filter logic - so perhaps an intermediate `BlockAndBlobsAndDataColumns` variant may make sense until we fully switch over to 1D PeerDAS.

dapplion replied:
Can you 
```
match self {
  RpcBlockInner::*=> RpcBlock
}
```
so the enum type guarantees that blobs and columns do not exist at the same time

jimmygchen replied:
Not sure if I'm understanding correctly but `self` is actually a struct type which contains optional blobs and columns:

https://github.com/sigp/lighthouse/blob/e707536fd5468c58493b5185e6566cee9bf9d392/beacon_node/beacon_chain/src/data_availability_checker.rs#L603-L608

jimmygchen replied:
For an `RpcBlock`, we will have either `block + blobs` or `block + samples` if we've transitioned to DAS.

eserilev commented on beacon_node/beacon_chain/src/beacon_chain.rs:
```diff
@@ -3212,6 +3245,23 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
     }
 
+    /// Checks if the provided data column can make any cached blocks available, and imports immediately
+    /// if so, otherwise caches the blob in the data availability checker.
```
```suggestion
    /// if so, otherwise caches the data column in the data availability checker.
```

eserilev commented on beacon_node/beacon_chain/src/block_verification_types.rs:
```diff
@@ -140,19 +146,29 @@ impl<E: EthSpec> RpcBlock<E> {
     pub fn n_blobs(&self) -> usize {
         match &self.block {
-            RpcBlockInner::Block(_) => 0,
+            RpcBlockInner::Block(_) | RpcBlockInner::BlockAndDataColumns(_, _) => 0,
```
I think match statements are executed in order from left to right. There might be a weird edge case here where both data columns and blobs are avail, but we return 0. If we're trying to support both blobs and data_columns initially, this edge case has the potential to cause issues

jimmygchen replied:
Yep, you're right - perhaps we can add the `BlockAndBlobsAndDataColumns` variant to the other arm?

jimmygchen replied:
So this type is used for RPC responses, so i don't think blobs and data column will coexist - we either request for `block + blobs` or `block + data columns` during lookups and sync, depending on the fork we're at.

eserilev commented on beacon_node/beacon_chain/src/data_availability_checker.rs:
```diff
@@ -242,35 +269,45 @@ impl<T: BeaconChainTypes> DataAvailabilityChecker<T> {
-                };
+            (maybe_blob_list, maybe_data_column_list) => {
+                let (verified_blobs, verified_data_column) =
+                    if self.blobs_required_for_block(&block) {
```
is there a similar `data_columns_required_for_blocks` check we'll need to make here in the future?

jimmygchen replied:
They are the same condition - `blobs_required_for_block` checks for number of commitments in the block - if there are commitments, there are blobs and we should verify data columns if they exist.

eserilev commented on beacon_node/beacon_chain/src/data_column_verification.rs:
```diff
@@ -0,0 +1,218 @@
+    ///
+    /// ## Peer scoring
+    ///
+    /// We were unable to process this blob due to an internal error. It's
```
```suggestion
    /// We were unable to process this data column due to an internal error. It's
```

eserilev commented on beacon_node/beacon_chain/src/historical_blocks.rs:
```diff
@@ -127,6 +128,12 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
+            if let Some(data_columns) = maybe_data_columns {
+                // new_oldest_data_column_slot = Some(block.slot());
+                self.store
+                    .data_columns_as_kv_store_ops(&block_root, data_columns, &mut blob_batch);
```
I think we may want a `Vec<KeyValueStoreOp>` dedicated to data_columns instead of re-using `blob_batch`

jimmygchen replied:
I think the benefit I see with using a dedicated one is if we want to write to a separate database? I'm re-using the blob db for simplicity here (since the maintainence/pruning logic are the same), but happy to create a separate db if there's good reasons to do that.

jimmygchen replied:
So I'm reusing the same blobs db for data columns but using a different `DbColumn` to separate data columns from blobs.

It feels like we could benefit from better UX - no change for users that's already using the `blobs-db` flag and save on some code duplication.

eserilev commented on beacon_node/store/src/hot_cold_store.rs:
```diff
@@ -1448,6 +1524,32 @@ impl<E: EthSpec, Hot: ItemStore<E>, Cold: ItemStore<E>> HotColdDB<E, Hot, Cold>
         }
     }
 
+    /// Fetch blobs for a given block from the store.
```
```suggestion
    /// Fetch data_columns for a given block from the store.
```

eserilev commented on consensus/types/src/data_column_sidecar.rs:
```diff
@@ -17,6 +24,15 @@ pub type ColumnIndex = u64;
 pub type Cell<T> = FixedVector<u8, <T as EthSpec>::FieldElementsPerCell>;
 pub type DataColumn<T> = VariableList<Cell<T>, <T as EthSpec>::MaxBlobsPerBlock>;
 
+/// Container of the data that identifies an individual blob.
```
```suggestion
/// Container of the data that identifies an individual data column.
```

jimmygchen committed:
Apply suggestions from code review

Co-authored-by: Eitan Seri-Levi <eserilev@gmail.com>
Co-authored-by: Jacob Kaufmann <jacobkaufmann18@gmail.com>

jimmygchen committed:
Merge branch 'das' into data-cols-by-root

# Conflicts:
#	beacon_node/beacon_chain/src/data_availability_checker.rs
#	beacon_node/beacon_chain/src/data_availability_checker/availability_view.rs
#	consensus/types/src/chain_spec.rs

jimmygchen committed:
Fix build.

jimmygchen committed:
Store `DataColumnInfo` in database and various cleanups.

jimmygchen commented:
@dapplion @eserilev 
I've addressed the review comments and I think this is ready for another review. 

There are a few things that need more work but I think it's probably best to keep them separate from this one as this PR is already quite bloated (mostly boilerplates) and further work depends on some of the changes here. It would be good to merge this soon and iterate from there.

There are a few things to note that are not covered here:
1. There's no guarantee that the data column will be stored properly, because:
    - We don't wait for them to arrive before importing the block - I think we'll need another condition under [`chain.process_gossip_data_column`](https://github.com/jimmygchen/lighthouse/blob/e707536fd5468c58493b5185e6566cee9bf9d392/beacon_node/beacon_chain/src/beacon_chain.rs#L2940) to handle the new code path, i.e. if block has been imported, just verify and store them on disk, otherwise it goes through the existing DA checker code path.
    - Look up via RPC has not yet been implemented
2. There are some recent changes to the `consensus-spec` PeerDAS PR, I think it's best to update them in a new PR.

AgeManning commented on beacon_node/network/src/router.rs:
```diff
@@ -637,6 +644,17 @@ impl<T: BeaconChainTypes> Router<T> {
+        _data_column_sidecar: Option<Arc<DataColumnSidecar<T::EthSpec>>>,
+    ) {
+        // TODO(das) implement `DataColumnsByRoot` response handling
+        unimplemented!()
```
Before we merge this, perhaps we want to just leave a comment here rather than panicing?

jimmygchen replied:
Updated in https://github.com/sigp/lighthouse/pull/5196/commits/3a967847ece611ee5486b4061a8b13434b383258

AgeManning commented on consensus/types/src/data_column_sidecar.rs:
```diff
@@ -121,13 +137,62 @@ impl<T: EthSpec> DataColumnSidecar<T> {
+            column: VariableList::new(vec![Cell::<T>::default(); T::MaxBlobsPerBlock::to_usize()])
+                .unwrap(),
+            kzg_commitments: VariableList::new(vec![
+                KzgCommitment::empty_for_testing();
```
The sizes the same here? Empty and non-empty?

jimmygchen replied:
The "empty" `KzgCommitment` and `KzgProof` means a fixed size array filled with `0`s. So here it's filling the `Vec`s with 4096 (`MaxBlobCommitmentsPerBlock`) instances of empty `KzgCommitment`s and `KzgProof`s (0 filled fixed-size array).

Although I think we won't actually have `MaxBlobCommitmentsPerBlock` (4096) number of items here - this upper bound used in the type was to make future upgrades easier but in practice this would be bound to `MaxBlobsPerBlock` (6) instead. I'll update this.

jimmygchen replied:
Updated in https://github.com/sigp/lighthouse/pull/5196/commits/3a967847ece611ee5486b4061a8b13434b383258

AgeManning reviewed (approved):
I skimmed this. Looks like a lot of boiler plate. Others may want to review, but happy to merge down to build on this.

jimmygchen committed:
Update `DataColumnSidecar` ssz max size and remove panic code.
